src/ui/ui_settings/ui_settings_service.js

```diff
@@ -0,0 +1,130 @@
+import { defaultsDeep, noop } from 'lodash';
+import { errors as esErrors } from 'elasticsearch';
+
+import { getDefaultSettings } from './defaults';
+
+function hydrateUserSettings(userSettings) {
+  return Object.keys(userSettings)
+    .map(key => ({ key, userValue: userSettings[key] }))
+    .filter(({ userValue }) => userValue !== null)
+    .reduce((acc, { key, userValue }) => ({ ...acc, [key]: { userValue } }), {});
+}
+
+export class UiSettingsService {
+  constructor(options = {}) {
+    const {
+      index,
+      type,
+      id,
+      callCluster,
+      readInterceptor = noop,
+    } = options;
+
+    this._callCluster = callCluster;
+    this._readInterceptor = readInterceptor;
+    this._index = index;
+    this._type = type;
+    this._id = id;
+  }
+
+  getDefaults() {
+    return getDefaultSettings();
+  }
+
+  // returns a Promise for the value of the requested setting
+  async get(key) {
+    const all = await this.getAll();
+    return all[key];
+  }
+
+  async getAll() {
+    const raw = await this.getRaw();
+
+    return Object.keys(raw)
+      .reduce((all, key) => {
+        const item = raw[key];
+        const hasUserValue = 'userValue' in item;
+        all[key] = hasUserValue ? item.userValue : item.value;
+        return all;
+      }, {});
+  }
+
+  async getRaw() {
+    const userProvided = await this.getUserProvided();
+    return defaultsDeep(userProvided, this.getDefaults());
+  }
+
+  async getUserProvided(options) {
+    return hydrateUserSettings(await this._read(options));
+  }
+
+  async setMany(changes) {
+    await this._write(changes);
+  }
+
+  async set(key, value) {
+    await this.setMany({ [key]: value });
+  }
+
+  async remove(key) {
+    await this.set(key, null);
+  }
+
+  async removeMany(keys) {
+    const changes = {};
+    keys.forEach(key => {
+      changes[key] = null;
+    });
+    await this.setMany(changes);
+  }
+
+  async _write(changes) {
+    await this._callCluster('update', {
+      index: this._index,
+      type: this._type,
+      id: this._id,
+      body: {
+        doc: changes
+      }
+    });
+  }
+
+  async _read(options = {}) {
+    const interceptValue = await this._readInterceptor(options);
+    if (interceptValue != null) {
```

Comment:
> I usually prefer specific checks, but I'm okey with != null in JS (with a typesystem it should be more specific). I just read it as "I expect x to be defined"

Yeah, it's arguable. Let's just use a single source of truth. Currently Kibana says [one thing](https://github.com/elastic/kibana/blob/master/style_guides/js_style_guide.md#use-the--operator), but ESLint config [makes an exception for `null`](https://github.com/elastic/eslint-config-kibana/blob/master/.eslintrc.js#L34). If we rely on ESLint config, I'll just make a note and update Kibana Style Guide later on (once I have enough edits).

> (in JS some people set a value to null to mean not defined, while others do undefined — until we are more consistent on that, I think != null makes sense)

Sure, can we agree on that here and later propose it to our style guide?